File: samples/source/go/hello.go
package main

import (
    "fmt"
    "os"
    "runtime"
    "sync"
    "time"
)

// Global variables
var (
    globalCounter int = 42
    globalMutex   sync.Mutex
)

// Interface example
type Speaker interface {
    Speak() string
}

// Struct with methods
type Application struct {
    Name    string
    Version string
    Debug   bool
}

func (a *Application) Speak() string {
    return fmt.Sprintf("%s v%s", a.Name, a.Version)
}

func (a *Application) String() string {
    return a.Speak()
}

// Goroutine function
func worker(id int, jobs <-chan int, results chan<- int, wg *sync.WaitGroup) {
    defer wg.Done()
    for job := range jobs {
        time.Sleep(10 * time.Millisecond)
        results <- job * 2
    }
}

// Function with defer and panic recovery
func riskyOperation() (result int, err error) {
    defer func() {
        if r := recover(); r != nil {
            err = fmt.Errorf("recovered from panic: %v", r)
        }
    }()
    
    // Simulate some work
    result = 100
    
    // Intentional panic for testing
    if false {
        panic("simulated panic")
    }
    
    return result, nil
}

// Generic function (Go 1.18+)
func mapSlice[T any, U any](slice []T, fn func(T) U) []U {
    result := make([]U, len(slice))
    for i, v := range slice {
        result[i] = fn(v)
    }
    return result
}

func main() {
    fmt.Println("Hello, World from Go!")
    fmt.Printf("Runtime: %s/%s\n", runtime.GOOS, runtime.GOARCH)
    fmt.Printf("Go version: %s\n", runtime.Version())
    
    // Struct usage
    app := &Application{
        Name:    "glaurung",
        Version: "1.0.0",
        Debug:   true,
    }
    fmt.Printf("Application: %s\n", app)
    
    // Channel and goroutine example
    numJobs := 5
    jobs := make(chan int, numJobs)
    results := make(chan int, numJobs)
    
    var wg sync.WaitGroup
    wg.Add(3)
    
    // Start workers
    for w := 1; w <= 3; w++ {
        go worker(w, jobs, results, &wg)
    }
    
    // Send jobs
    for j := 1; j <= numJobs; j++ {
        jobs <- j
    }
    close(jobs)
    
    // Wait for completion in background
    go func() {
        wg.Wait()
        close(results)
    }()
    
    // Collect results
    fmt.Print("Results: ")
    for result := range results {
        fmt.Printf("%d ", result)
    }
    fmt.Println()
    
    // Map/slice operations with generics
    numbers := []int{1, 2, 3, 4, 5}
    doubled := mapSlice(numbers, func(n int) int { return n * 2 })
    fmt.Printf("Doubled: %v\n", doubled)
    
    // Error handling
    if value, err := riskyOperation(); err != nil {
        fmt.Printf("Error: %v\n", err)
    } else {
        fmt.Printf("Operation result: %d\n", value)
    }
    
    // Command line arguments
    if len(os.Args) > 1 {
        fmt.Printf("Arguments: %v\n", os.Args[1:])
    }
    
    os.Exit(0)
}

File: samples/source/rust/hello.rs
// Rust sample with various language features for binary analysis

use std::collections::HashMap;
use std::env;
use std::sync::{Arc, Mutex};
use std::thread;
use std::time::Duration;

// Global constant
const GLOBAL_CONSTANT: i32 = 42;

// Static mutable (unsafe)
static mut GLOBAL_COUNTER: i32 = 0;

// Trait definition
trait Speaker {
    fn speak(&self) -> String;
}

// Struct with generics
#[derive(Debug, Clone)]
struct Application<T> {
    name: String,
    version: String,
    data: T,
}

// Trait implementation
impl<T: std::fmt::Debug> Speaker for Application<T> {
    fn speak(&self) -> String {
        format!("{} v{} with data: {:?}", self.name, self.version, self.data)
    }
}

// Enum with pattern matching
#[derive(Debug)]
enum Operation {
    Add(i32, i32),
    Multiply(i32, i32),
    Divide(i32, i32),
    Print(String),
}

// Function with Result type
fn risky_operation(value: i32) -> Result<i32, String> {
    if value == 0 {
        Err("Cannot process zero".to_string())
    } else if value < 0 {
        Err("Negative values not supported".to_string())
    } else {
        Ok(value * 2)
    }
}

// Async-like function with threads
fn parallel_computation(values: Vec<i32>) -> Vec<i32> {
    let shared_results = Arc::new(Mutex::new(Vec::new()));
    let mut handles = vec![];
    
    for value in values {
        let results = Arc::clone(&shared_results);
        let handle = thread::spawn(move || {
            // Simulate work
            thread::sleep(Duration::from_millis(10));
            let computed = value * value;
            let mut res = results.lock().unwrap();
            res.push(computed);
        });
        handles.push(handle);
    }
    
    for handle in handles {
        handle.join().unwrap();
    }
    
    let results = shared_results.lock().unwrap();
    results.clone()
}

// Macro definition
macro_rules! debug_print {
    ($($arg:tt)*) => {
        #[cfg(debug_assertions)]
        println!("[DEBUG] {}", format!($($arg)*));
    };
}

// Function with iterator chains
fn process_data(data: &[i32]) -> i32 {
    data.iter()
        .filter(|&&x| x > 0)
        .map(|&x| x * 2)
        .fold(0, |acc, x| acc + x)
}

// Closure example
fn create_multiplier(factor: i32) -> impl Fn(i32) -> i32 {
    move |x| x * factor
}

fn main() {
    println!("Hello, World from Rust!");
    
    // Command line arguments
    let args: Vec<String> = env::args().collect();
    if args.len() > 1 {
        println!("Arguments: {:?}", &args[1..]);
    }
    
    // Struct usage with generics
    let app = Application {
        name: String::from("glaurung"),
        version: String::from("1.0.0"),
        data: HashMap::from([
            ("feature1", true),
            ("feature2", false),
            ("feature3", true),
        ]),
    };
    println!("Application: {}", app.speak());
    
    // Pattern matching
    let operations = vec![
        Operation::Add(10, 20),
        Operation::Multiply(5, 7),
        Operation::Divide(100, 5),
        Operation::Print(String::from("Processing complete")),
    ];
    
    for op in operations {
        match op {
            Operation::Add(a, b) => println!("Add: {} + {} = {}", a, b, a + b),
            Operation::Multiply(a, b) => println!("Multiply: {} × {} = {}", a, b, a * b),
            Operation::Divide(a, b) => {
                if b != 0 {
                    println!("Divide: {} ÷ {} = {}", a, b, a / b);
                } else {
                    println!("Division by zero!");
                }
            }
            Operation::Print(msg) => println!("Message: {}", msg),
        }
    }
    
    // Error handling with Result
    let test_values = vec![10, 0, -5, 42];
    for val in test_values {
        match risky_operation(val) {
            Ok(result) => println!("Operation({}) = {}", val, result),
            Err(e) => println!("Error for {}: {}", val, e),
        }
    }
    
    // Thread-based parallel computation
    let numbers = vec![1, 2, 3, 4, 5];
    let results = parallel_computation(numbers.clone());
    println!("Parallel computation results: {:?}", results);
    
    // Iterator chains
    let data = vec![1, -2, 3, -4, 5, 6];
    let sum = process_data(&data);
    println!("Processed sum: {}", sum);
    
    // Closure usage
    let multiply_by_3 = create_multiplier(3);
    println!("3 × 7 = {}", multiply_by_3(7));
    
    // Unsafe block
    unsafe {
        GLOBAL_COUNTER += 1;
        println!("Global counter: {}", GLOBAL_COUNTER);
    }
    
    // Macro usage
    debug_print!("Program completed successfully");
    
    // String manipulation
    let mut text = String::from("Rust");
    text.push_str(" is memory safe!");
    println!("{}", text);
    
    // Option type
    let maybe_number: Option<i32> = Some(42);
    if let Some(n) = maybe_number {
        println!("The answer is {}", n);
    }
    
    std::process::exit(0);
}

// Module with tests
#[cfg(test)]
mod tests {
    use super::*;
    
    #[test]
    fn test_risky_operation() {
        assert!(risky_operation(10).is_ok());
        assert!(risky_operation(0).is_err());
        assert!(risky_operation(-5).is_err());
    }
    
    #[test]
    fn test_process_data() {
        let data = vec![1, -2, 3, -4, 5];
        assert_eq!(process_data(&data), 18); // (1+3+5)*2
    }
}